File: mm/memory_reservations.c
#include "memory_reservations.h"

#include <stddef.h>

#include "mm_constants.h"
#include "../lib/klog.h"
#include "../drivers/serial.h"
#include "../lib/alignment.h"
#include "../boot/kernel_panic.h"
#include "../lib/memory.h"
#include "../lib/string.h"
#include "mm_constants.h"

/* Fallback storage if caller does not provide a buffer. */
#define MM_REGION_STATIC_CAP 1024
static mm_region_t static_region_store[MM_REGION_STATIC_CAP];

typedef struct mm_region_store {
    mm_region_t *regions;
    uint32_t capacity;
    uint32_t count;
    uint32_t overflows;
    int configured;
} mm_region_store_t;

static mm_region_store_t region_store = {
    .regions = static_region_store,
    .capacity = MM_REGION_STATIC_CAP,
    .count = 0,
    .overflows = 0,
    .configured = 0,
};

static void copy_label(char dest[32], const char *src) {
    if (!src) {
        dest[0] = '\0';
        return;
    }

    size_t i = 0;
    for (; i < 31 && src[i] != '\0'; i++) {
        dest[i] = src[i];
    }
    dest[i] = '\0';
}

static void ensure_storage(void) {
    if (!region_store.regions || region_store.capacity == 0) {
        kernel_panic("MM: region storage not configured");
    }
}

static void clear_region(mm_region_t *region) {
    region->phys_base = 0;
    region->length = 0;
    region->kind = MM_REGION_RESERVED;
    region->type = MM_RESERVATION_ALLOCATOR_METADATA;
    region->flags = 0;
    region->label[0] = '\0';
}

static void clear_store(void) {
    ensure_storage();
    for (uint32_t i = 0; i < region_store.capacity; i++) {
        clear_region(&region_store.regions[i]);
    }
    region_store.count = 0;
    region_store.overflows = 0;
}

void mm_region_map_configure(mm_region_t *buffer, uint32_t capacity) {
    if (!buffer || capacity == 0) {
        kernel_panic("MM: invalid region storage configuration");
    }
    region_store.regions = buffer;
    region_store.capacity = capacity;
    region_store.configured = 1;
    clear_store();
}

void mm_region_map_reset(void) {
    if (!region_store.configured) {
        /* Switch to fallback storage on first reset if user did not configure. */
        region_store.regions = static_region_store;
        region_store.capacity = MM_REGION_STATIC_CAP;
        region_store.configured = 1;
    }
    clear_store();
}

static void insert_slot(uint32_t index) {
    ensure_storage();
    if (region_store.count >= region_store.capacity) {
        region_store.overflows++;
        kernel_panic("MM: region map capacity exceeded");
    }

    if (index > region_store.count) {
        index = region_store.count;
    }

    if (region_store.count > 0 && index < region_store.count) {
        memmove(&region_store.regions[index + 1],
                &region_store.regions[index],
                (region_store.count - index) * sizeof(mm_region_t));
    }
    region_store.count++;
    clear_region(&region_store.regions[index]);
}

static int regions_equivalent(const mm_region_t *a, const mm_region_t *b) {
    if (!a || !b) {
        return 0;
    }
    if (a->kind != b->kind) {
        return 0;
    }
    if (a->kind == MM_REGION_USABLE) {
        return a->flags == b->flags && a->label[0] == b->label[0];
    }
    return a->type == b->type &&
           a->flags == b->flags &&
           strncmp(a->label, b->label, sizeof(a->label)) == 0;
}

static void try_merge_with_neighbors(uint32_t index) {
    if (region_store.count == 0 || index >= region_store.count) {
        return;
    }

    /* Merge with previous if adjacent and equivalent */
    if (index > 0) {
        mm_region_t *curr = &region_store.regions[index];
        mm_region_t *prev = &region_store.regions[index - 1];
        uint64_t prev_end = prev->phys_base + prev->length;
        if (prev_end == curr->phys_base && regions_equivalent(prev, curr)) {
            prev->length += curr->length;
            memmove(curr, curr + 1,
                    (region_store.count - index - 1) * sizeof(mm_region_t));
            region_store.count--;
            if (index > 0) {
                index--;
            }
        }
    }

    /* Merge with next if adjacent and equivalent */
    if (index + 1 < region_store.count) {
        mm_region_t *curr = &region_store.regions[index];
        mm_region_t *next = &region_store.regions[index + 1];
        uint64_t curr_end = curr->phys_base + curr->length;
        if (curr_end == next->phys_base && regions_equivalent(curr, next)) {
            curr->length += next->length;
            memmove(next, next + 1,
                    (region_store.count - index - 2) * sizeof(mm_region_t));
            region_store.count--;
        }
    }
}

static uint32_t find_region_index(uint64_t phys_base) {
    uint32_t idx = 0;
    while (idx < region_store.count &&
           region_store.regions[idx].phys_base + region_store.regions[idx].length <= phys_base) {
        idx++;
    }
    return idx;
}

static void split_region(uint32_t index, uint64_t split_base) {
    if (index >= region_store.count) {
        return;
    }
    mm_region_t *region = &region_store.regions[index];
    uint64_t region_end = region->phys_base + region->length;
    if (split_base <= region->phys_base || split_base >= region_end) {
        return;
    }

    insert_slot(index + 1);
    mm_region_t *right = &region_store.regions[index + 1];
    *right = *region;
    right->phys_base = split_base;
    right->length = region_end - split_base;
    region->length = split_base - region->phys_base;
}

static void overlay_region(uint64_t phys_base, uint64_t length,
                           mm_region_kind_t kind, mm_reservation_type_t type,
                           uint32_t flags, const char *label) {
    if (length == 0) {
        return;
    }

    /* Reject obvious virtual/HHDM addresses that are not physical. */
    if (phys_base >= KERNEL_VIRTUAL_BASE || phys_base >= HHDM_VIRT_BASE) {
        klog_printf(KLOG_INFO, "MM: rejecting virtual overlay base 0x%llx\n",
                    (unsigned long long)phys_base);
        kernel_panic("MM: region overlay received virtual address");
    }

    uint64_t end = phys_base + length;
    if (end <= phys_base) {
        kernel_panic("MM: region overlay overflow");
    }

    uint64_t aligned_base = align_down_u64(phys_base, PAGE_SIZE_4KB);
    uint64_t aligned_end = align_up_u64(end, PAGE_SIZE_4KB);
    if (aligned_end <= aligned_base) {
        kernel_panic("MM: region overlay collapsed");
    }

    uint64_t cursor = aligned_base;
    while (cursor < aligned_end) {
        uint32_t idx = find_region_index(cursor);

        if (idx >= region_store.count || region_store.regions[idx].phys_base > cursor) {
            /* Insert new gap before existing region or append at end. */
            insert_slot(idx);
            region_store.regions[idx].phys_base = cursor;
            region_store.regions[idx].length = aligned_end - cursor;
            region_store.regions[idx].kind = kind;
            region_store.regions[idx].type = type;
            region_store.regions[idx].flags = flags;
            copy_label(region_store.regions[idx].label, label);
            try_merge_with_neighbors(idx);
            break;
        }

        mm_region_t *region = &region_store.regions[idx];
        uint64_t region_end = region->phys_base + region->length;

        /* Split so that [cursor, aligned_end) aligns with region boundaries. */
        split_region(idx, cursor);
        region = &region_store.regions[idx];
        region_end = region->phys_base + region->length;

        uint64_t apply_end = (aligned_end < region_end) ? aligned_end : region_end;
        split_region(idx, apply_end);
        region = &region_store.regions[idx];

        /* Overwrite region slice with new attributes. */
        region->kind = kind;
        region->type = type;
        region->flags = flags;
        copy_label(region->label, label);
        try_merge_with_neighbors(idx);

        cursor = apply_end;
    }
}

int mm_region_add_usable(uint64_t phys_base, uint64_t length, const char *label) {
    if (length == 0) {
        return -1;
    }
    overlay_region(phys_base, length, MM_REGION_USABLE, MM_RESERVATION_FIRMWARE_OTHER, 0, label);
    return 0;
}

int mm_region_reserve(uint64_t phys_base, uint64_t length,
                      mm_reservation_type_t type, uint32_t flags,
                      const char *label) {
    if (length == 0) {
        return -1;
    }
    overlay_region(phys_base, length, MM_REGION_RESERVED, type, flags, label);
    return 0;
}

/* Debug helper: emit all regions with physical ranges. */
void mm_region_dump(enum klog_level level) {
    for (uint32_t i = 0; i < region_store.count; i++) {
        const mm_region_t *region = &region_store.regions[i];
        const char *kind = (region->kind == MM_REGION_USABLE) ? "usable" : "reserved";
        klog_printf(level, "[MM] %s: 0x%llx - 0x%llx (%llu KB) label=%s flags=0x%x\n",
                    kind,
                    (unsigned long long)region->phys_base,
                    (unsigned long long)(region->phys_base + region->length - 1),
                    (unsigned long long)(region->length / 1024),
                    region->label[0] ? region->label : "-",
                    region->flags);
    }
}

uint32_t mm_region_count(void) {
    return region_store.count;
}

const mm_region_t *mm_region_get(uint32_t index) {
    if (index >= region_store.count) {
        return NULL;
    }
    return &region_store.regions[index];
}

uint32_t mm_reservations_count(void) {
    uint32_t count = 0;
    for (uint32_t i = 0; i < region_store.count; i++) {
        if (region_store.regions[i].kind == MM_REGION_RESERVED &&
            region_store.regions[i].length > 0) {
            count++;
        }
    }
    return count;
}

uint32_t mm_reservations_capacity(void) {
    return region_store.capacity;
}

uint32_t mm_reservations_overflow_count(void) {
    return region_store.overflows;
}

const mm_region_t *mm_reservations_get(uint32_t index) {
    /* Iterate deterministic by physical order; skip unusable. */
    uint32_t seen = 0;
    for (uint32_t i = 0; i < region_store.count; i++) {
        const mm_region_t *region = &region_store.regions[i];
        if (region->kind != MM_REGION_RESERVED || region->length == 0) {
            continue;
        }
        if (seen == index) {
            return region;
        }
        seen++;
    }
    return NULL;
}

const mm_region_t *mm_reservations_find(uint64_t phys_addr) {
    for (uint32_t i = 0; i < region_store.count; i++) {
        const mm_region_t *region = &region_store.regions[i];
        if (region->kind != MM_REGION_RESERVED || region->length == 0) {
            continue;
        }
        uint64_t end = region->phys_base + region->length;
        if (phys_addr >= region->phys_base && phys_addr < end) {
            return region;
        }
    }
    return NULL;
}

int mm_is_reserved(uint64_t phys_addr) {
    return mm_reservations_find(phys_addr) != NULL;
}

int mm_is_range_reserved(uint64_t phys_base, uint64_t length) {
    if (length == 0) {
        return 0;
    }

    uint64_t end = phys_base + length;
    if (end <= phys_base) {
        return 1;
    }

    for (uint32_t i = 0; i < region_store.count; i++) {
        const mm_region_t *region = &region_store.regions[i];
        if (region->kind != MM_REGION_RESERVED || region->length == 0) {
            continue;
        }
        uint64_t region_end = region->phys_base + region->length;
        if (region->phys_base < end && region_end > phys_base) {
            return 1;
        }
    }

    return 0;
}

void mm_iterate_reserved(mm_region_iter_cb cb, void *ctx) {
    if (!cb) {
        return;
    }

    for (uint32_t i = 0; i < region_store.count; i++) {
        const mm_region_t *region = &region_store.regions[i];
        if (region->kind != MM_REGION_RESERVED || region->length == 0) {
            continue;
        }
        cb(region, ctx);
    }
}

const char *mm_reservation_type_name(mm_reservation_type_t type) {
    switch (type) {
        case MM_RESERVATION_ALLOCATOR_METADATA:
            return "allocator metadata";
        case MM_RESERVATION_FRAMEBUFFER:
            return "framebuffer";
        case MM_RESERVATION_ACPI_RECLAIMABLE:
            return "acpi reclaim";
        case MM_RESERVATION_ACPI_NVS:
            return "acpi nvs";
        case MM_RESERVATION_APIC:
            return "apic";
        case MM_RESERVATION_FIRMWARE_OTHER:
            return "firmware";
        default:
            return "reserved";
    }
}

uint64_t mm_reservations_total_bytes(uint32_t required_flags) {
    uint64_t total = 0;
    for (uint32_t i = 0; i < region_store.count; i++) {
        const mm_region_t *region = &region_store.regions[i];
        if (region->kind != MM_REGION_RESERVED || region->length == 0) {
            continue;
        }
        if (required_flags != 0 && (region->flags & required_flags) != required_flags) {
            continue;
        }
        total += region->length;
    }
    return total;
}

uint64_t mm_region_total_bytes(mm_region_kind_t kind) {
    uint64_t total = 0;
    for (uint32_t i = 0; i < region_store.count; i++) {
        if (region_store.regions[i].kind == kind) {
            total += region_store.regions[i].length;
        }
    }
    return total;
}

uint64_t mm_region_highest_usable_frame(void) {
    uint64_t highest = 0;
    for (uint32_t i = 0; i < region_store.count; i++) {
        const mm_region_t *region = &region_store.regions[i];
        if (region->kind != MM_REGION_USABLE || region->length == 0) {
            continue;
        }
        uint64_t end = region->phys_base + region->length - 1;
        uint64_t frame = end >> 12;
        if (frame > highest) {
            highest = frame;
        }
    }
    return highest;
}

File: mm/src/memory_reservations.rs
use core::ffi::{c_char, c_int};

use slopos_ostd::sync::{LOCK_LEVEL_RESOURCE, SpinLock};
use slopos_ostd::util::cstr::cstr_from_kernel_ptr;
use slopos_utils::{align_down_u64, align_up_u64, klog_info};

use crate::memory_layout_defs::KERNEL_VIRTUAL_BASE;
use crate::paging_defs::PAGE_SIZE_4KB;

const MM_REGION_STATIC_CAP: usize = 4096;

pub const MM_RESERVATION_FLAG_EXCLUDE_ALLOCATORS: u32 = 1 << 0;
pub const MM_RESERVATION_FLAG_ALLOW_MM_PHYS_TO_VIRT: u32 = 1 << 1;
pub const MM_RESERVATION_FLAG_MMIO: u32 = 1 << 2;

#[repr(C)]
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum MmReservationType {
    AllocatorMetadata = 0,
    Framebuffer = 1,
    AcpiReclaimable = 2,
    AcpiNvs = 3,
    Apic = 4,
    FirmwareOther = 5,
}

#[repr(C)]
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum MmRegionKind {
    Usable = 0,
    Reserved = 1,
}

#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct MmRegion {
    pub phys_base: u64,
    pub length: u64,
    pub kind: MmRegionKind,
    pub type_: MmReservationType,
    pub flags: u32,
    pub label: [u8; 32],
}

impl MmRegion {
    pub const fn zeroed() -> Self {
        Self {
            phys_base: 0,
            length: 0,
            kind: MmRegionKind::Reserved,
            type_: MmReservationType::AllocatorMetadata,
            flags: 0,
            label: [0; 32],
        }
    }
}

struct RegionStoreInner {
    regions: [MmRegion; MM_REGION_STATIC_CAP],
    count: u32,
    overflows: u32,
}

impl RegionStoreInner {
    const fn capacity(&self) -> u32 {
        MM_REGION_STATIC_CAP as u32
    }
}

static REGION_STORE: SpinLock<RegionStoreInner> = SpinLock::new(
    RegionStoreInner {
        regions: [MmRegion::zeroed(); MM_REGION_STATIC_CAP],
        count: 0,
        overflows: 0,
    },
    LOCK_LEVEL_RESOURCE,
);

fn with_store<R>(f: impl FnOnce(&mut RegionStoreInner) -> R) -> R {
    let mut guard = REGION_STORE.lock();
    f(&mut guard)
}

fn clear_store_inner(store: &mut RegionStoreInner) {
    for slot in store.regions.iter_mut() {
        *slot = MmRegion::zeroed();
    }
    store.count = 0;
    store.overflows = 0;
}

fn copy_label(dest: &mut [u8; 32], src: *const c_char) {
    let Some(bytes) = cstr_from_kernel_ptr(src) else {
        dest[0] = 0;
        return;
    };
    let take = bytes.len().min(31);
    dest[..take].copy_from_slice(&bytes[..take]);
    dest[take] = 0;
}

fn insert_slot(store: &mut RegionStoreInner, index: u32) -> Result<(), ()> {
    if store.count >= store.capacity() {
        store.overflows = store.overflows.saturating_add(1);
        return Err(());
    }

    let cap = store.count as usize;
    let idx = (index as usize).min(cap);
    if cap > 0 && idx < cap {
        // shift `[idx, count)` one slot to the right.
        store.regions.copy_within(idx..cap, idx + 1);
    }
    store.count += 1;
    store.regions[idx] = MmRegion::zeroed();
    Ok(())
}

fn regions_equivalent(a: &MmRegion, b: &MmRegion) -> bool {
    if a.kind != b.kind {
        return false;
    }
    if matches!(a.kind, MmRegionKind::Usable) {
        a.flags == b.flags && a.label[0] == b.label[0]
    } else {
        a.type_ == b.type_ && a.flags == b.flags && a.label == b.label
    }
}

fn try_merge_with_neighbors(store: &mut RegionStoreInner, index: u32) {
    let count = store.count;
    if count == 0 || index >= count {
        return;
    }
    let i = index as usize;

    // Merge with previous.
    if index > 0 {
        let prev_end = store.regions[i - 1].phys_base + store.regions[i - 1].length;
        let merge = prev_end == store.regions[i].phys_base && {
            let prev = store.regions[i - 1];
            let curr = store.regions[i];
            regions_equivalent(&prev, &curr)
        };
        if merge {
            store.regions[i - 1].length = store.regions[i - 1]
                .length
                .wrapping_add(store.regions[i].length);
            let cap = store.count as usize;
            if i + 1 < cap {
                store.regions.copy_within(i + 1..cap, i);
            }
            store.count -= 1;
        }
    }

    // Merge with next (re-read count after possible previous merge).
    let count = store.count;
    if (index + 1) < count {
        let i = index as usize;
        let curr_end = store.regions[i].phys_base + store.regions[i].length;
        let merge = curr_end == store.regions[i + 1].phys_base && {
            let curr = store.regions[i];
            let next = store.regions[i + 1];
            regions_equivalent(&curr, &next)
        };
        if merge {
            store.regions[i].length = store.regions[i]
                .length
                .wrapping_add(store.regions[i + 1].length);
            let cap = store.count as usize;
            if i + 2 < cap {
                store.regions.copy_within(i + 2..cap, i + 1);
            }
            store.count -= 1;
        }
    }
}

fn find_region_index(store: &RegionStoreInner, phys_base: u64) -> u32 {
    let mut idx = 0u32;
    while idx < store.count {
        let region = &store.regions[idx as usize];
        if region.phys_base + region.length > phys_base {
            break;
        }
        idx += 1;
    }
    idx
}

fn split_region(store: &mut RegionStoreInner, index: u32, split_base: u64) -> Result<(), ()> {
    if index >= store.count {
        return Err(());
    }
    let i = index as usize;
    let region = store.regions[i];
    let region_end = region.phys_base + region.length;
    if split_base <= region.phys_base || split_base >= region_end {
        return Ok(());
    }

    insert_slot(store, index + 1)?;
    let i = index as usize;
    let mut right = region;
    right.phys_base = split_base;
    right.length = region_end - split_base;
    store.regions[i + 1] = right;
    store.regions[i].length = split_base - store.regions[i].phys_base;
    Ok(())
}

fn overlay_region(
    phys_base: u64,
    length: u64,
    kind: MmRegionKind,
    type_: MmReservationType,
    flags: u32,
    label: *const c_char,
) -> c_int {
    if length == 0 {
        return -1;
    }

    if phys_base >= KERNEL_VIRTUAL_BASE {
        klog_info!("MM: rejecting virtual overlay base 0x{:x}", phys_base);
        return -1;
    }
    if crate::hhdm::is_available() && phys_base >= crate::hhdm::offset() {
        klog_info!("MM: rejecting virtual overlay base 0x{:x}", phys_base);
        return -1;
    }

    let end = phys_base.wrapping_add(length);
    if end <= phys_base {
        return -1;
    }

    let aligned_base = align_down_u64(phys_base, PAGE_SIZE_4KB);
    let aligned_end = align_up_u64(end, PAGE_SIZE_4KB);
    if aligned_end <= aligned_base {
        return -1;
    }

    with_store(|store| {
        let mut cursor = aligned_base;
        while cursor < aligned_end {
            let idx = find_region_index(store, cursor);

            let region_exists = idx < store.count;
            let needs_insert = !region_exists || store.regions[idx as usize].phys_base > cursor;
            if needs_insert {
                if insert_slot(store, idx).is_err() {
                    return -1;
                }
                let i = idx as usize;
                let region = &mut store.regions[i];
                region.phys_base = cursor;
                region.length = aligned_end - cursor;
                region.kind = kind;
                region.type_ = type_;
                region.flags = flags;
                copy_label(&mut region.label, label);
                try_merge_with_neighbors(store, idx);
                break;
            }

            if split_region(store, idx, cursor).is_err() {
                return -1;
            }
            let i = idx as usize;
            let region_end = store.regions[i].phys_base + store.regions[i].length;
            let apply_end = if aligned_end < region_end {
                aligned_end
            } else {
                region_end
            };
            if split_region(store, idx, apply_end).is_err() {
                return -1;
            }

            let i = idx as usize;
            let region = &mut store.regions[i];
            region.kind = kind;
            region.type_ = type_;
            region.flags = flags;
            copy_label(&mut region.label, label);
            try_merge_with_neighbors(store, idx);

            cursor = apply_end;
        }
        0
    })
}

pub fn mm_region_map_reset() {
    with_store(clear_store_inner);
}

pub fn mm_region_add_usable(phys_base: u64, length: u64, label: *const c_char) -> c_int {
    if length == 0 {
        return -1;
    }
    overlay_region(
        phys_base,
        length,
        MmRegionKind::Usable,
        MmReservationType::FirmwareOther,
        0,
        label,
    )
}

pub fn mm_region_reserve(
    phys_base: u64,
    length: u64,
    type_: MmReservationType,
    flags: u32,
    label: *const c_char,
) -> c_int {
    if length == 0 {
        return -1;
    }
    overlay_region(
        phys_base,
        length,
        MmRegionKind::Reserved,
        type_,
        flags,
        label,
    )
}

pub fn mm_region_count() -> u32 {
    with_store(|store| store.count)
}

pub fn mm_region_get(index: u32) -> Option<MmRegion> {
    with_store(|store| {
        if index >= store.count {
            None
        } else {
            Some(store.regions[index as usize])
        }
    })
}

pub fn mm_reservations_count() -> u32 {
    with_store(|store| {
        let mut count = 0;
        for i in 0..store.count as usize {
            let region = &store.regions[i];
            if matches!(region.kind, MmRegionKind::Reserved) && region.length > 0 {
                count += 1;
            }
        }
        count
    })
}

pub fn mm_reservations_capacity() -> u32 {
    with_store(|store| store.capacity())
}

pub fn mm_reservations_overflow_count() -> u32 {
    with_store(|store| store.overflows)
}

pub fn mm_reservations_get(index: u32) -> Option<MmRegion> {
    with_store(|store| {
        let mut seen = 0;
        for i in 0..store.count as usize {
            let region = &store.regions[i];
            if !matches!(region.kind, MmRegionKind::Reserved) || region.length == 0 {
                continue;
            }
            if seen == index {
                return Some(*region);
            }
            seen += 1;
        }
        None
    })
}

pub fn mm_reservations_find(phys_addr: u64) -> Option<MmRegion> {
    with_store(|store| {
        for i in 0..store.count as usize {
            let region = &store.regions[i];
            if !matches!(region.kind, MmRegionKind::Reserved) || region.length == 0 {
                continue;
            }
            let end = region.phys_base + region.length;
            if phys_addr >= region.phys_base && phys_addr < end {
                return Some(*region);
            }
        }
        None
    })
}

pub fn mm_reservations_find_option(phys_addr: u64) -> Option<MmRegion> {
    mm_reservations_find(phys_addr)
}

pub fn mm_reservation_type_name(type_: MmReservationType) -> &'static str {
    match type_ {
        MmReservationType::AllocatorMetadata => "allocator metadata",
        MmReservationType::Framebuffer => "framebuffer",
        MmReservationType::AcpiReclaimable => "acpi reclaim",
        MmReservationType::AcpiNvs => "acpi nvs",
        MmReservationType::Apic => "apic",
        MmReservationType::FirmwareOther => "firmware",
    }
}

pub fn mm_reservations_total_bytes(required_flags: u32) -> u64 {
    with_store(|store| {
        let mut total = 0u64;
        for i in 0..store.count as usize {
            let region = &store.regions[i];
            if !matches!(region.kind, MmRegionKind::Reserved) || region.length == 0 {
                continue;
            }
            if required_flags != 0 && (region.flags & required_flags) != required_flags {
                continue;
            }
            total = total.wrapping_add(region.length);
        }
        total
    })
}

pub fn mm_region_total_bytes(kind: MmRegionKind) -> u64 {
    with_store(|store| {
        let mut total = 0u64;
        for i in 0..store.count as usize {
            let region = &store.regions[i];
            if region.kind == kind {
                total = total.wrapping_add(region.length);
            }
        }
        total
    })
}

pub fn mm_region_highest_usable_frame() -> u64 {
    with_store(|store| {
        let mut highest = 0u64;
        for i in 0..store.count as usize {
            let region = &store.regions[i];
            if !matches!(region.kind, MmRegionKind::Usable) || region.length == 0 {
                continue;
            }
            let end = region.phys_base + region.length - 1;
            let frame = end >> 12;
            if frame > highest {
                highest = frame;
            }
        }
        highest
    })
}

/// Returns the highest frame index seen across **any** memory-map region
/// (Usable, Reserved, KernelAndModules, Bootloader-Reclaimable,
/// AcpiReclaimable, AcpiNvs, BadMemory, …). The OSTD `META_SLOTS` array
/// must cover every paddr the kernel might wrap with `Frame<M>` —
/// including the kernel image, the bootloader-allocated PML4, and any
/// framebuffer / ACPI region that gets mapped via cursor. The "usable"
/// variant above is too narrow for that.
pub fn mm_region_highest_frame_seen() -> u64 {
    with_store(|store| {
        let mut highest = 0u64;
        for i in 0..store.count as usize {
            let region = &store.regions[i];
            if region.length == 0 {
                continue;
            }
            let end = region.phys_base.saturating_add(region.length - 1);
            let frame = end >> 12;
            if frame > highest {
                highest = frame;
            }
        }
        highest
    })
}
